### PR TITLE
02 Mask R-CNN notebook draft

### PR DESCRIPTION
**Pull request by simonzhaoms**
### Description

Add 02 Mask R-CNN notebook draft.  It's not fully complete yet.  Just for some comments on the existing contents.  Lack of evaluation metrics demo and detailed explanations for the contents.

### Checklist:

- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by PatrickBue** [comment]
@simonzhaoms I just left a few high-level point in the notebook itself. Maybe you were already thinking about addressing these points.
@jiata Let's discuss at some point a common plotting function which plots the image, and ground truth, and maybe also prediction. Also also plots segmentation masks/keypoints if provided. This might be too much all in a single plotting function - let's discuss. 
Also, let's think about if we want to just have a single class "DetectionClass" or a separate one for segmentation.

**Comment by jiata** [comment]
Super cool to see this notebook here! Also, the additional functionality you added to the bboxes are very cool and useful.

Before diving into the notebook with individual comments, i have a few higher-level questions that i think needs to be addressed.

1. Do we want to use the pedestrian data set over the fridge objects?
  I think it might actually be preferable to use the fridge objects so that we can clearly show how we're using a pretrained model first, then fine-tuning on a custom dataset. This will also give us the opportunity to talk about the annotation format a little bit more since we'll have created it ourselves.

1. Should we be making new Dataset and Learner classes? 
  I haven't thought it through entirely, but i think we should be able to add more functionality to the `DetectionDataset` and `DetectionLearner` to accommodate segmentation instead of making entirely new classes. Do you see a problem with this way of doing it instead? Right now, it actually seems like the dataset/learner classes you created are supersets of the original `DetectionDatset`/`DetectionLearner`. So, conceptually speaking, it seems like they should be the same class. Even some kind of inheritence structure could make sense, where `MaskLearner` and `KeyPointLearner` inherits from `DetectionLearner`. Or something like that. 

1. Keypoint feels like a different topic and i'm wondering if we should move it to a separate notebook. 
  My feeling generally is that each notebook should address 1 scenario and essentially act as an executable for that scenario. This is essentially what we do with all our other notebooks. The option of using keypoint in this notebook breaks from this tradition. (small side note, can you add the reference urls for where you got the keypoint.py information from or if you copied methods over from another place?)

1. I think your notebook could do a bit more in terms of holding the users hands through understanding the annotation requirements. 
  Because we're introducing a new annotation system here, we should assume that users know the bare minimum. I think this point also relates to the first point above about creating our own dataset for maskrcnn.

**Comment by simonzhaoms** [comment]
> Super cool to see this notebook here! Also, the additional functionality you added to the bboxes are very cool and useful.
> 
> Before diving into the notebook with individual comments, i have a few higher-level questions that i think needs to be addressed.
> 
> 1. Do we want to use the pedestrian data set over the fridge objects?
>    I think it might actually be preferable to use the fridge objects so that we can clearly show how we're using a pretrained model first, then fine-tuning on a custom dataset. This will also give us the opportunity to talk about the annotation format a little bit more since we'll have created it ourselves.
> 2. Should we be making new Dataset and Learner classes?
>    I haven't thought it through entirely, but i think we should be able to add more functionality to the `DetectionDataset` and `DetectionLearner` to accommodate segmentation instead of making entirely new classes. Do you see a problem with this way of doing it instead? Right now, it actually seems like the dataset/learner classes you created are supersets of the original `DetectionDatset`/`DetectionLearner`. So, conceptually speaking, it seems like they should be the same class. Even some kind of inheritence structure could make sense, where `MaskLearner` and `KeyPointLearner` inherits from `DetectionLearner`. Or something like that.
> 3. Keypoint feels like a different topic and i'm wondering if we should move it to a separate notebook.
>    My feeling generally is that each notebook should address 1 scenario and essentially act as an executable for that scenario. This is essentially what we do with all our other notebooks. The option of using keypoint in this notebook breaks from this tradition. (small side note, can you add the reference urls for where you got the keypoint.py information from or if you copied methods over from another place?)
> 4. I think your notebook could do a bit more in terms of holding the users hands through understanding the annotation requirements.
>    Because we're introducing a new annotation system here, we should assume that users know the bare minimum. I think this point also relates to the first point above about creating our own dataset for maskrcnn.

1. Agree, and it will make it more easy and natural for users to follow.  Do you know any tools I can use for annotating segmentation/masks?  In addition, we still need to find some dataset for keypoints since only person keypoints are supported in the PyTorch pretrained model.

2. Why I use inheritance is because I am afraid of changing too much of the existing code would introduce bugs for other contents and code depend on it, though I have changed a lot, :joy: I will figure it out.

3. Agree.  (Good point :+1: I will add references in `keypoint.py`)

4. Yes.  I will add that these days.

**Comment by simonzhaoms** [comment]
So the next steps:
1. Add annotations for masks for fridge objects
2. For keypoints, what about the COCO 2017 Train/Val annotations in [COCO dataets](http://cocodataset.org/#download)?
3. Merge MaskLearner and KeypointLearner into DetectionLearner
4. Add references for `keypoint.py`
5. Tell a good story in the notebook :smile:

**Comment by PatrickBue** [comment]
@simonzhaoms I am hoping to speak with JS soon to address some of the questions he asked above and come up with recommendations for you. He's sick today, but hopefully back tomorrow.

**Comment by jiata** [comment]
Not sure if you're at a point where you want more feedback, but i noticed that the keypoint stuff is still in this PR. Could we move all key point code into a different PR? (including the util funcs and the part of the notebook that address the key point)

**Comment by PatrickBue** [comment]
@simonzhaoms Please try to make progress quickly on this PR. I feel that this Mask R-CNN work has been going on for quite a long time already. I will be on a similar time zone than you if you have questions

**Comment by simonzhaoms** [comment]
@PatrickBue @jiata I updated the code and the notebook with limited 31 annotated images from odFridgeObjects, removed the keypoints code and merged in the latest changes in the staging branch.  Feel free to leave comments.  

BTW, I am not entirely sure how the `allow_negatives` in `DetectionDataset.add_images()` and `DetectionDataset._read_annos()` work, so the added mask code may not be correct.  Please help me check, thanks.

File: utils_cv/detection/keypoint.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jiata) [inline utils_cv/detection/keypoint.py]: looks like the keypoint code is still here. can you put this in a separate PR?

File: utils_cv/detection/bbox.py [hunk 2]
```diff
@@ -1,7 +1,11 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-from typing import List, Union
+from .mask import binarise_mask
```
Comment (jiata) [inline utils_cv/detection/bbox.py]: [minor] i think the convention we've been using has to been to put references to our own lib at the end of the import statements.

File: utils_cv/detection/bbox.py [hunk 3]
```diff
@@ -25,15 +29,54 @@ def __init__(self, left: int, top: int, right: int, bottom: int):
         self.standardize()
 
     @classmethod
-    def from_array(cls, arr: List[int]) -> "Bbox":
+    def from_array(cls, arr: List[int]) -> "_Bbox":
```
Comment (jiata) [inline utils_cv/detection/bbox.py]: thanks for catching these

File: utils_cv/detection/bbox.py [hunk 4]
```diff
@@ -25,15 +29,54 @@ def __init__(self, left: int, top: int, right: int, bottom: int):
...
         return _Bbox(arr[0], arr[1], arr[2], arr[3])
 
     @classmethod
-    def from_array_xywh(cls, arr: List[int]) -> "Bbox":
+    def from_array_xywh(cls, arr: List[int]) -> "_Bbox":
         """ create a Bbox object from an array [left, top, width, height] """
         return _Bbox(arr[0], arr[1], arr[0] + arr[2], arr[1] + arr[3])
 
+    @classmethod
+    def get_rect_from_binary_mask(cls, binary_mask: np.ndarray) -> List[int]:
```
Comment (jiata) [inline utils_cv/detection/bbox.py]: [minor] can we add a description?

File: utils_cv/detection/plot.py [hunk 5]
```diff
@@ -84,19 +92,109 @@ def plot_boxes(
...
+    representing different objects, 0 as background.
+    """
+    if isinstance(im, (str, Path)):
+        im = Image.open(im)
+
+    # convert to RGBA for transparentising
+    im = im.convert('RGBA')
+    binary_masks = binarise_mask(mask)
+
+    # # allow single color as palette to be used for all instance
```
Comment (jiata) [inline utils_cv/detection/plot.py]: can we remove this bit that commented out?

File: utils_cv/detection/plot.py [hunk 6]
```diff
@@ -84,19 +92,109 @@ def plot_boxes(
...
+    ]
+    # merge masks into img one by one
+    for cmask in colored_masks:
+        tmask = Image.fromarray(transparentise_mask(cmask, plot_settings.mask_alpha))
+        im = Image.alpha_composite(im, tmask)
+
+    return im
+
+
+def plot_keypoints(
```
Comment (jiata) [inline utils_cv/detection/plot.py]: can we remove keypoint code into another Pr for now?

File: utils_cv/detection/plot.py [hunk 7]
```diff
@@ -84,19 +92,109 @@ def plot_boxes(
...
+            draw.line(
+                line,
+                fill=plot_settings.keypoint_color,
+                width=plot_settings.keypoint_th
+            )
+
+    return im
+
+
+def display_image(
```
Comment (jiata) [inline utils_cv/detection/plot.py]: good idea to separate this out, thanks!

File: utils_cv/detection/plot.py [hunk 8]
```diff
@@ -110,16 +208,39 @@ def display_bboxes(
...
+    """ Draw image with bounding boxes and mask. """
+    im = Image.open(im_path)
+    if mask_path is not None:
+        im = plot_mask(im_path, mask_path)
+    if bboxes is not None:
+        im = plot_boxes(im, bboxes)
+    display_image(im, ax)
+
+
+def display_bbox_mask_keypoint(
```
Comment (jiata) [inline utils_cv/detection/plot.py]: can we remove keypoint code and put it in another PR for now?

File: utils_cv/detection/model.py [hunk 9]
```diff
@@ -130,6 +128,113 @@ def get_pretrained_fasterrcnn(
...
 
+def _tune_mask_predictor(model: nn.Module, num_classes: int) -> nn.Module:
+    # get the number of input features of mask predictor from the pretrained model
+    in_features = model.roi_heads.mask_predictor.conv5_mask.in_channels
+    # replace the mask predictor with a new one
+    model.roi_heads.mask_predictor = MaskRCNNPredictor(in_features, 256, num_classes)
+    return model
+
+
+def _tune_keypoint_predictor(model: nn.Module, num_keypoints: int) -> nn.Module:
```
Comment (jiata) [inline utils_cv/detection/model.py]: move this to the keypoint pr

File: utils_cv/detection/model.py [hunk 10]
```diff
@@ -152,18 +257,43 @@ def _calculate_ap(e: CocoEvaluator) -> float:
...
-        self.model = model
+        self.device = device
+        if self.device is None:
+            self.device = torch_device()
+
         self.dataset = dataset
 
+        if dataset and "labels" in dataset.__dict__:
+            self.labels = ["__background__"] + dataset.labels
+        else:
```
Comment (jiata) [inline utils_cv/detection/model.py]: not sure if we should default to coco labels if not specified

File: utils_cv/detection/model.py [hunk 11]
```diff
@@ -247,18 +380,105 @@ def plot_precision_loss_curves(
...
+
+        res = self._get_det_bboxes(pred, im_path)
+        if "masks" in pred:
+            res = (res, pred["masks"].squeeze())
+        elif "keypoints" in pred:
+            res = (res, pred["keypoints"])
+        return res
+
+    @classmethod
+    def _pack_pred_results(cls, res: List, infos: Dict) -> List[Dict]:
```
Comment (jiata) [inline utils_cv/detection/model.py]: can you add docstring stating what this function does?

**Comment by jiata** [comment]
@simonzhaoms - why are the tests not triggering? i suspect it may be because you're making the PR from your own forked version. Can you try making a new branch in the CVBP repo and making the PR from there?

From scanning your code, im worried that your changes will break a lot of what is already in the other notebooks.

**Comment by simonzhaoms** [comment]
@jiata I am so sorry, my fault to waste your time for reviewing the old code, I forgot to push my changes onto GitHub yesterday.

I will add tests and docs today to make it better.

**Comment by simonzhaoms** [comment]
@jiata The mask data can be downloaded from https://github.com/simonzhaoms/testdata/archive/master.zip, but right now only 31 images.

**Comment by jiata** [comment]
ah ok! no wonder :P i'll take a look. Looks like the tests are getting triggered now. At the very least, lets get the linux-cpu/gpu tests passing. The windows tests are failing somewhat sporadically, so i wouldn't worry too much about it.

**Comment by jiata** [comment]
@simonzhaoms do you know why tests aren't running on this pr?

**Comment by simonzhaoms** [comment]
> @simonzhaoms do you know why tests aren't running on this pr?

I think that's because of merge conflicts.

**Comment by simonzhaoms** [comment]
> it looks like the linux gpu test didn't kick off

Not sure.  I didn't change anything related to testing pipeline.  Any suggestions that I can follow to figure it out?

**Comment by simonzhaoms** [comment]
> it looks like the linux gpu test didn't kick off

I think it is because this PR is from my fork not directly from this repo.  I created another PR https://github.com/microsoft/ComputerVision/pull/389 on a branch of this repo instead of my fork.  It triggered all testing pipelines.

**Comment by jiata** [comment]
> > it looks like the linux gpu test didn't kick off
> 
> I think it is because this PR is from my fork not directly from this repo. I created another PR #389 on a branch of this repo instead of my fork. It triggered all testing pipelines.

Thanks, i'll take a look at the new one

**Comment by simonzhaoms** [comment]
Close this PR, since it doesn't trigger all testing pipelines.  Use https://github.com/microsoft/ComputerVision/pull/389 instead.

File: utils_cv/detection/model.py (hunk 10 above)
Comment (simonzhaoms) [reply utils_cv/detection/model.py]: Done.  Saw you added `add_labels()` method.

File: utils_cv/detection/model.py (hunk 11 above)
Comment (simonzhaoms) [reply utils_cv/detection/model.py]: Done
